app/code/Magento/Bundle/Test/Unit/Observer/AppendUpsellProductsObserverTest.php

```diff
@@ -0,0 +1,411 @@
+     * @var Observer|MockObject
+     */
+    private $observerMock;
+
+    /**
+     * @var ProductLinkCollection|MockObject
+     */
+    private $collectionMock;
+
+    /**
+     * @var Product|MockObject
+     */
+    private $productMock;
+
+    /**
+     * @var Visibility|MockObject
+     */
+    private $productVisibilityMock;
+
+    /**
+     * @inheritdoc
+     */
+    protected function setUp() : void
+    {
+        $this->objectManager = new ObjectManager($this);
+        $this->observerMock = $this->createMock(Observer::class);
+
+        $this->bundleCollectionMock = $this->getMockBuilder(ProductCollection::class)
+            ->disableOriginalConstructor()
+            ->setMethods([
+                'addAttributeToSelect',
+                'addFieldToFilter',
+                'addFinalPrice',
+                'addMinimalPrice',
+                'addStoreFilter',
+                'addTaxPercents',
+                'load',
+                'setFlag',
+                'setPageSize',
+                'setVisibility'
+            ])
+            ->getMock();
+
+        $this->bundleDataMock = $this->getMockBuilder(BundleHelper::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getAllowedSelectionTypes'])
+            ->getMock();
+
+        $this->bundleSelectionMock = $this->getMockBuilder(Selection::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getParentIdsByChild'])
+            ->getMock();
+
+        $this->configMock = $this->getMockBuilder(CatalogConfig::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getProductAttributes'])
+            ->getMock();
+
+        $this->eventMock = $this->getMockBuilder(Event::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getProduct', 'getCollection', 'getLimit'])
+            ->getMock();
+
+        $this->collectionMock = $this->getMockBuilder(Event::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['setItems', 'getItems'])
+            ->getMock();
+
+        $this->productMock = $this->getMockBuilder(Product::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getCollection', 'getId', 'getTypeId'])
+            ->getMock();
+
+        $this->productVisibilityMock = $this->getMockBuilder(Visibility::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getVisibleInCatalogIds'])
+            ->getMock();
+
+        $this->observer = $this->objectManager->getObject(
+            AppendUpsellProductsObserver::class,
+            [
+                'bundleData' => $this->bundleDataMock,
+                'productVisibility' => $this->productVisibilityMock,
+                'config' => $this->configMock,
+                'bundleSelection' => $this->bundleSelectionMock,
+            ]
+        );
+    }
+
+    /**
+     * Test observer execute method
+     */
+    public function testExecute()
```

Comment:
Rename to reflect what is given and what are the assertions.

Eg. "testExecuteForXLoadsCollectionOnce"

Reply:
sure :)

Thanks, @lbajsarowicz

Reply:
**done**